chat-plugins/r9k.js

```diff
@@ -0,0 +1,157 @@
+/*
+* Robot9000 mode chat plugin.
+*
+* For an explanation on what Robot9000 is, see <https://blog.xkcd.com/2008/01/14/robot9000-and-xkcd-signal-attacking-noise-in-chat/>
+* This implementation of r9k mode is a variation of the original idea, trying to focus more on filtering often spammed messages, with a much smaller load on disk space and long-term RAM usage.
+* Each chat message is converted into a sha1 checksum, which is then put in a temporary cache. When a message is repeated, and therefore triggers the r9k filter, it is moved into a permanent storage.
+*
+* Written by bumbadadabum.
+*/
+
+'use strict';
+
+const FS = require('../fs');
+const crypto = require('crypto');
+
+const R9K_DIRECTORY = ('config/r9k');
+
+class R9KCache {
+	constructor() {
+		this.caches = new Map();
+	}
+
+	init(room, checksums) {
+		this.caches.set(room.id, new Set(checksums));
+	}
+
+	add(room, checksum) {
+		if (!this.caches.has(room.id)) this.caches.set(room.id, new Set());
+		this.caches.get(room.id).add(checksum);
+	}
+
+	has(room, checksum) {
+		let cache = this.caches.get(room.id);
+		return cache && cache.has(checksum);
+	}
+}
+
+const tempCache = new R9KCache();
+const permanentCache = new R9KCache();
+
+function sha1(str) {
+	 let generator = crypto.createHash('sha1');
+	 generator.update(str);
+	 return generator.digest('hex');
+}
+
+function checkR9k(user, room, message) {
+	let checksum = sha1(toId(message));
+
+	if (tempCache.has(room, checksum) || permanentCache.has(room, checksum)) {
+		if (room.chatRoomData && !permanentCache.has(room, checksum)) {
+			appendPermanent(room, checksum);
+		}
+
+		if (user.can('mute', null, room)) return;
+
+		if (!room.r9kPunishments) room.r9kPunishments = new Map();
+		let userid = toId(user);
+
+		let muteTime = room.r9kPunishments.get(userid);
+
+		if (!muteTime) {
+			room.r9kPunishments.set(userid, 2);
+			return 2;
+		}
+
+		muteTime *= 2;
+		room.r9kPunishments.set(userid, muteTime);
+		return muteTime;
+	} else {
+		tempCache.add(room, checksum);
+		return 0;
+	}
+}
+
+/**
+ * @param {Room} room
+ * @param {string} checksum
+ */
+async function loadPermanent(room) {
+	let data = await FS(`${R9K_DIRECTORY}/${room.id}.txt`).readTextIfExists();
+	data = data.split("\n").filter(val => val.length === 40);
+	if (data.length) permanentCache.init(room, data);
+}
+
+/**
+ * @param {Room} room
+ * @param {string} checksum
+ */
+async function appendPermanent(room, checksum) {
+	permanentCache.add(room, checksum);
+
+	await prune();
+
+	FS(`${R9K_DIRECTORY}/${room.id}.txt`).append(`${checksum}\n`);
+}
+
+async function prune() {
+	let files = await FS(R9K_DIRECTORY).readdir();
+	files = files.filter(file => file.endsWith('.txt') && !Rooms(file.split('.')[0]));
+	for (let i = 0; i < files.length; i++) {
+		FS(`${R9K_DIRECTORY}/${files[i]}`).unlinkIfExists();
+	}
+}
+
+module.exports = {
+	chatfilter: function (message, user, room) {
+		if (room.r9k && !user.can('bypassall')) {
```

Comment:
``room && room.r9k``

``room`` is not always defined, i.e, if someone sends a PM

Reply:
i'm pretty sure that it doesnt matter since there is no "!chatfilter" alias which allows the command to be used in PMs

Reply:
This isn't a command. It's the new chatfilter plugin feature.